### PR TITLE
Allow arithmetic in enum definitions

### PR DESCRIPTION
**Pull request by kerams**
Expanding on #14370 and https://github.com/fsharp/fslang-suggestions/issues/539.

~~One minor problem are error messages. Consider cases with inconsistent types:~~

b9Ye9eyW7X

RatAEk93di

~~In the latter case, `TcExpr` seems to insert type casts of some sort (I think), which prevent `EvalLiteralExprOrAttribArg` from evaluating the expression, so type checking fails before enum case type consistency could be checked and we don't get the nice message in the first picture. I could use `TcExprOfUnknownType`, but I don't know how to "bind" the returned type to the type variable `fieldTy`.~~
Sorted

Also, the parentheses are required, but I think that's fine.

### CODE REVIEW

**Comment by T-Gro** [comment]
Can you please add a test for the diagnostics just like in the screenshot?

**Comment by T-Gro** [comment]
The () will be required only in the enum case, right?
Just to verify - would be good to know if Fantomas keeps those, otherwise a change would be needed there as well.

**Comment by kerams** [comment]
> The () will be required only in the enum case, right?

Sure, otherwise everything would be broken :). I've only swapped enum case constants for expressions in the parser.

**Comment by kerams** [comment]
All good now, I think.

File: src/Compiler/Checking/CheckDeclarations.fs [hunk 1]
```diff
@@ -563,26 +563,38 @@ module TcRecdUnionAndEnumDeclarations =
...
-            let checkXmlDocs = cenv.diagnosticOptions.CheckXmlDocs
-            let xmlDoc = xmldoc.ToXmlDoc(checkXmlDocs, Some [])
-            Construct.NewRecdField true (Some v) id false thisTy false false [] attrs xmlDoc vis false
-      
-    let TcEnumDecls (cenv: cenv) env parent thisTy enumCases =
+            TcEnumCaseConstDecl cenv env parent attrs thisTy m id xmldoc v
+        | _ ->
+            if cenv.g.langVersion.SupportsFeature LanguageFeature.ArithmeticInLiterals then
+                let expr, actualTy, _ = TcExprOfUnknownType cenv env tpenv v
+                UnifyTypes cenv env m fieldTy actualTy
```
Comment (kerams) [inline src/Compiler/Checking/CheckDeclarations.fs:586]: Bit of a sidenote, the signature `val UnifyTypes: cenv: TcFileState -> env: TcEnv -> m: range -> actualTy: TType -> expectedTy: TType -> unit` might be incorrect - the last parameter seems to be used as the actual type.

Comment (kerams) [reply src/Compiler/Checking/CheckDeclarations.fs:586]: I'm confused by what constitutes an expected and actual type.

Checking `if` expressions - unit is the expected `then` branch type when there is no `else`

```fsharp
if not isRecovery && Option.isNone synElseExprOpt then
     UnifyTypes cenv env m g.unit_ty overallTy.Commit
```

Checking `do` expressions - unit is the expected time.

```fsharp
| SynExpr.Do (synInnerExpr, m) ->
    UnifyTypes cenv env m overallTy.Commit g.unit_ty
```

**Comment by kerams** [comment]
I could also take this as an opportunity adjust error ranges if you want. https://sharplab.io/#v2:DYLgZgzgNAJiDUAfALgTwA4FMAEBBbAvALABQ252i2AGodgIwAypFl2AmnQExA== should only highlight the expression range.

File: tests/service/TreeVisitorTests.fs [hunk 2]
```diff
@@ -57,5 +57,5 @@ let ``Visit enum definition test`` () =
     let parseTree = parseSourceCode("C:\\test.fs", source)
 
     match SyntaxTraversal.Traverse(pos0, parseTree, visitor) with
-    | Some [ SynEnumCase (_, SynIdent(id1,_), _, _, _, _, _); SynEnumCase (_, SynIdent(id2,_), _, _, _, _, _) ] when id1.idText = "A" && id2.idText = "B" -> ()
+    | Some [ SynEnumCase (_, SynIdent(id1,_), _, _, _, _); SynEnumCase (_, SynIdent(id2,_), _, _, _, _) ] when id1.idText = "A" && id2.idText = "B" -> ()
```
Comment (T-Gro) [inline tests/service/TreeVisitorTests.fs:60]:
```suggestion
    | Some [ SynEnumCase (ident=SynIdent(id1,_)); SynEnumCase (ident=SynIdent(id2,_)) ] when id1.idText = "A" && id2.idText = "B" -> ()
```

**Comment by T-Gro** [comment]
> I could also take this as an opportunity to adjust error ranges if you want. https://sharplab.io/#v2:DYLgZgzgNAJiDUAfALgTwA4FMAEBBbAvALABQ252i2AGodgIwAypFl2AmnQExA== should only highlight the expression range.


Agree, it makes more sense to use the expression's range only.

**Review by T-Gro — APPROVED** [review]

**Review by T-Gro — DISMISSED** [review]

**Comment by kerams** [comment]
Bump

File: src/Compiler/Checking/CheckDeclarations.fs (hunk 1 above)
Comment (T-Gro) [reply src/Compiler/Checking/CheckDeclarations.fs]: The g.unity_ty is the expected one, that matches also my understanding of expected vs. actual. 
The usages seem to confirm this view, with only a few outliers => BUT, changing things around now can lead to different (inverted) error message. So it should be first checked if the changed meaning is better.

image

**Review by T-Gro — APPROVED** [review]

**Review by vzarytovskii — APPROVED** [review]

**Pull request merged**